File: csharp/Original/PublishingHouse/Books/Authors/AuthorIdOrData.cs
using PublishingHouse.Books.Entities;
using PublishingHouse.Core.Validation;

namespace PublishingHouse.Books.Authors;

public record AuthorIdOrData
{
    public AuthorId? AuthorId { get; }
    public AuthorFirstName? FirstName { get; }
    public AuthorLastName? LastName { get; }

    public AuthorIdOrData(AuthorId? authorId, AuthorFirstName? firstName, AuthorLastName? lastName)
    {
        if (authorId == null)
        {
            firstName.AssertNotNull();
            lastName.AssertNotNull();
        }

        AuthorId = authorId;
        FirstName = firstName;
        LastName = lastName;
    }

    public (AuthorFirstName, AuthorLastName) Data => (FirstName.AssertNotNull(), LastName.AssertNotNull());
}

File: java/src/main/java/io/eventdriven/slimdownaggregates/original/domain/books/authors/AuthorIdOrData.java
package io.eventdriven.slimdownaggregates.original.domain.books.authors;

import io.eventdriven.slimdownaggregates.original.domain.books.entities.AuthorFirstName;
import io.eventdriven.slimdownaggregates.original.domain.books.entities.AuthorId;
import io.eventdriven.slimdownaggregates.original.domain.books.entities.AuthorLastName;

public record AuthorIdOrData(
  AuthorId authorId,
  AuthorFirstName firstName,
  AuthorLastName lastName) {
  public AuthorIdOrData {
    if ((authorId != null && (firstName == null || lastName == null)) ||
      (authorId == null && (firstName == null || lastName == null))) {
      throw new IllegalArgumentException("either authorId or first name and last name has to be provided");
    }
  }
}
